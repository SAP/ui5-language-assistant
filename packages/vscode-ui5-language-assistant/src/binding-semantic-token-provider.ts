import {
  TextDocument,
  SemanticTokens,
  SemanticTokensBuilder,
  DocumentSemanticTokensProvider,
  SemanticTokensLegend,
} from "vscode";
import { getSemanticTokens, tokenTypesLegend } from "./utils";

const tokenModifiersLegend = [];
export const bindingLegend = new SemanticTokensLegend(
  tokenTypesLegend,
  tokenModifiersLegend
);
class BindingSemanticTokensProvider implements DocumentSemanticTokensProvider {
  async provideDocumentSemanticTokens(
    document: TextDocument
  ): Promise<SemanticTokens> {
    const builder = new SemanticTokensBuilder();
    const documentUri = document.uri.toString();
    const semanticTokens = await getSemanticTokens({
      documentUri,
      content: document.getText(),
    });
    semanticTokens.forEach((item) =>
      builder.push(
        item.line,
        item.char,
        item.length,
        item.tokenType,
        item.tokenModifiers
      )
    );
    return builder.build();
  }
}

export const bindingSemanticTokensProvider =
  new BindingSemanticTokensProvider();
