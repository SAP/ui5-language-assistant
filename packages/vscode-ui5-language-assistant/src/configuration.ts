import {
  ConfigurationChangeEvent,
  Event,
  ExtensionContext,
  WorkspaceConfiguration,
} from "vscode";
import { LanguageClient } from "vscode-languageclient";

export const LOGGING_LEVEL_CONFIG_PROP = "UI5LanguageAssistant.logging.level";

export function listenToLogLevelChanges(opts: {
  context: ExtensionContext;
  client: LanguageClient;
  onDidChangeConfiguration: Event<ConfigurationChangeEvent>;
  getConfiguration: (section?: string) => WorkspaceConfiguration;
}): void {
  opts.context.subscriptions.push(
    opts.onDidChangeConfiguration((e) => {
      if (e.affectsConfiguration(LOGGING_LEVEL_CONFIG_PROP)) {
        const newLogLevel = opts
          .getConfiguration()
          .get(LOGGING_LEVEL_CONFIG_PROP);
        // no validation done here as the server performs such validation
        // and will ignore invalid values.
        opts.client.sendRequest("changeLogLevel", newLogLevel);
      }
    })
  );
}
