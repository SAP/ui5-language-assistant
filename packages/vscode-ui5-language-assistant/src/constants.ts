export const COMMAND_OPEN_DEMOKIT = "UI5LanguageAssistant.command.openDemokit";
export const LOGGING_LEVEL_CONFIG_PROP = "UI5LanguageAssistant.logging.level";
