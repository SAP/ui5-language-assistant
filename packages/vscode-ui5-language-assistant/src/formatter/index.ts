export { formatDocument, formatRange } from "./document";
