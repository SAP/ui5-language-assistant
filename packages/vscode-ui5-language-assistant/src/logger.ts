import {
  getLogger as logger,
  ILogger,
} from "@ui5-language-assistant/logic-utils";
import { join } from "path";
import { readJsonSync } from "fs-extra";

const getPackageName = (): string => {
  let meta;
  try {
    // in production - ../extensions/saposs.vscode-ui5-language-assistant-4.0.6/dist which equals to vscode-ui5-language-assistant/dist
    meta = readJsonSync(join(__dirname, "..", "package.json"));
  } catch (error) {
    meta = readJsonSync(join(__dirname, "..", "..", "package.json"));
  }
  return meta?.name;
};

const name = getPackageName();
export const getLogger = (): ILogger => {
  return logger(name);
};
