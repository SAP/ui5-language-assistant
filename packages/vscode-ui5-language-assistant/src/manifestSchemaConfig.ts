export interface IJSONValidationExtensionPoint {
  fileMatch: string[];
  comment?: string;
  url: string;
}

export const MANIFEST_FILE_MATCH = "manifest.json";
export const MANIFEST_COMMENT =
  "Automatic configuration for manifest.json schema - added by UI5 Language Assistant";

export function getManifestSchemaConfigEntry(): IJSONValidationExtensionPoint {
  const manifestSchemaConfig = {
    fileMatch: [MANIFEST_FILE_MATCH],
    comment: MANIFEST_COMMENT,
    url:
      "https://cdn.jsdelivr.net/gh/SAP/ui5-language-assistant/packages/vscode-ui5-language-assistant/resources/manifest-schema/rel-1.19/schema/schema.json",
  };

  return manifestSchemaConfig;
}
