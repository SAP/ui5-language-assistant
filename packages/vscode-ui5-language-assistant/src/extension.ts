/* istanbul ignore file */
import { resolve } from "path";
import { readFileSync } from "fs";
import {
  workspace,
  ExtensionContext,
  StatusBarItem,
  window,
  StatusBarAlignment,
  commands,
  env,
  Uri,
} from "vscode";
import {
  LanguageClient,
  LanguageClientOptions,
  ServerOptions,
  TransportKind,
} from "vscode-languageclient/node";
import { LogLevel } from "@vscode-logging/types";
import {
  SERVER_PATH,
  ServerInitializationOptions,
} from "@ui5-language-assistant/language-server";
import { COMMAND_OPEN_DEMOKIT, LOGGING_LEVEL_CONFIG_PROP } from "./constants";

type UI5Model = { url: string; framework: string; version: string };

let client: LanguageClient;
let statusBarItem: StatusBarItem;
let currentModel: UI5Model | undefined;

export async function activate(context: ExtensionContext): Promise<void> {
  // create the LanguageClient (+Server)
  client = createLanguageClient(context);

  // create the StatusBarItem which displays the used UI5 version
  statusBarItem = createStatusBarItem(context);

  // show/hide and update the status bar
  client.start().then(() => {
    client.onNotification("UI5LanguageAssistant/ui5Model", (model: UI5Model) =>
      updateCurrentModel(model)
    );
  });
  window.onDidChangeActiveTextEditor(() => {
    updateCurrentModel(undefined);
  });

  client.start();
}

function createLanguageClient(context: ExtensionContext): LanguageClient {
  const debugOptions = { execArgv: ["--nolazy", "--inspect=6009"] };

  const serverOptions: ServerOptions = {
    run: { module: SERVER_PATH, transport: TransportKind.ipc },
    debug: {
      module: SERVER_PATH,
      transport: TransportKind.ipc,
      options: debugOptions,
    },
  };

  const meta = JSON.parse(
    readFileSync(resolve(context.extensionPath, "package.json"), "utf8")
  );

  const logLevel = workspace.getConfiguration().get(LOGGING_LEVEL_CONFIG_PROP);

  const initializationOptions: ServerInitializationOptions = {
    modelCachePath: context.globalStoragePath,
    publisher: meta.publisher,
    name: meta.name,
    // validation of the logLevel value is done on the language server process.
    logLevel: logLevel as LogLevel,
  };

  const clientOptions: LanguageClientOptions = {
    documentSelector: [{ scheme: "file", pattern: "**/*.{view,fragment}.xml" }],
    synchronize: {
      fileEvents: [
        workspace.createFileSystemWatcher("**/manifest.json"),
        workspace.createFileSystemWatcher("**/ui5.yaml"),
        workspace.createFileSystemWatcher("**/*.xml"),
        workspace.createFileSystemWatcher("**/*.cds"),
        workspace.createFileSystemWatcher("**/package.json"),
      ],
    },
    outputChannelName: meta.displayName,
    initializationOptions: initializationOptions,
  };

  return new LanguageClient(
    "UI5LanguageAssistant",
    "UI5 Language Assistant",
    serverOptions,
    clientOptions
  );
}

function createStatusBarItem(context: ExtensionContext): StatusBarItem {
  // create and register the command to open the SAPUI5 demokit
  context.subscriptions.push(
    commands.registerCommand(COMMAND_OPEN_DEMOKIT, () => {
      env.openExternal(Uri.parse(`${currentModel?.url}`));
    })
  );

  // create a statusbar item to display the currently used UI5 version
  statusBarItem = window.createStatusBarItem(StatusBarAlignment.Right, 100);
  statusBarItem.command = COMMAND_OPEN_DEMOKIT;
  return statusBarItem;
}

function updateCurrentModel(model: UI5Model | undefined) {
  currentModel = model;
  if (statusBarItem) {
    if (currentModel) {
      statusBarItem.tooltip = `${currentModel.framework} Version (XMLView Editor)`;
      statusBarItem.text = `$(notebook-mimetype) ${currentModel.version}${
        currentModel.framework === "OpenUI5" ? "'" : ""
      }`;
      statusBarItem.show();
    } else {
      statusBarItem.hide();
    }
  }
}

export function deactivate(): Thenable<void> | undefined {
  if (!client) {
    return undefined;
  }
  updateCurrentModel(undefined);
  return client.stop();
}
