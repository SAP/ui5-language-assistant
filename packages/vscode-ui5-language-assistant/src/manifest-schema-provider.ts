import {
  Uri,
  TextDocumentContentProvider,
  EventEmitter,
  ExtensionContext,
} from "vscode";
import { tryFetch } from "@ui5-language-assistant/logic-utils";
import { SCHEMA_URI, MANIFEST_SCHEMA } from "./constants";
import { getSchemaContent } from "./utils";

class ManifestSchemaProvider implements TextDocumentContentProvider {
  private _schemaContent = "";
  set schemaContent(content: string) {
    this._schemaContent = content;
  }
  get schemaContent(): string {
    return this._schemaContent;
  }
  onDidChange() {
    return new EventEmitter<Uri>();
  }
  provideTextDocumentContent(uri: Uri): string {
    if (MANIFEST_SCHEMA === uri.scheme) {
      return this.schemaContent;
    }
    return "";
  }
}

const schemaProvider = new ManifestSchemaProvider();

/**
 * Get manifest schema provider. It first tries to load "schema.json" over internet. If it fails, it tries
 * to read it from local resource
 */
export const getManifestSchemaProvider = async (
  context: ExtensionContext
): Promise<ManifestSchemaProvider> => {
  let content = "";
  const response = await tryFetch(SCHEMA_URI);
  if (response) {
    content = await response.text();
  } else {
    content = await getSchemaContent(context);
  }
  schemaProvider.schemaContent = content;
  return schemaProvider;
};
