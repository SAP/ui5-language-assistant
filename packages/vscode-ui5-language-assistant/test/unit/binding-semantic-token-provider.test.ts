import { bindingSemanticTokensProvider } from "../../src/binding-semantic-token-provider";
import { TextDocument } from "vscode";
import {
  Config,
  ProjectName,
  ProjectType,
  TestFramework,
} from "@ui5-language-assistant/test-framework";

describe("binding-semantic-token-provider", () => {
  let testFramework: TestFramework;
  let documentUri;
  const viewFilePathSegments = [
    "app",
    "manage_travels",
    "webapp",
    "ext",
    "main",
    "Main.view.xml",
  ];
  beforeAll(function () {
    const useConfig: Config = {
      projectInfo: {
        name: ProjectName.cap,
        type: ProjectType.CAP,
        npmInstall: false,
      },
    };
    testFramework = new TestFramework(useConfig);
    documentUri = testFramework.getFileUri(viewFilePathSegments);
  });

  afterEach(() => {
    jest.restoreAllMocks();
  });
  it("test provideDocumentSemanticTokens", async () => {
    const text = `
<mvc:View
      xmlns="sap.m"
      xmlns:mvc="sap.ui.core.mvc">
      <Text id="test-id" text="{path: 'some-value'}"/>
</mvc:View>`;
    const documents = {
      uri: documentUri,
      getText: () => text,
    } as unknown as TextDocument;
    const result =
      await bindingSemanticTokensProvider.provideDocumentSemanticTokens(
        documents
      );
    expect(result).toMatchSnapshot();
  });
});
