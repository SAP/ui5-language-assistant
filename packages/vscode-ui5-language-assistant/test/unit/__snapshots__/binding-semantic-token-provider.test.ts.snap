// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`binding-semantic-token-provider test provideDocumentSemanticTokens 1`] = `
Array [
  Object {
    "char": 31,
    "length": 1,
    "line": 4,
    "tokenModifiers": undefined,
    "tokenType": 5,
  },
  Object {
    "char": 32,
    "length": 4,
    "line": 4,
    "tokenModifiers": undefined,
    "tokenType": 0,
  },
  Object {
    "char": 36,
    "length": 1,
    "line": 4,
    "tokenModifiers": undefined,
    "tokenType": 3,
  },
  Object {
    "char": 38,
    "length": 12,
    "line": 4,
    "tokenModifiers": undefined,
    "tokenType": 1,
  },
  Object {
    "char": 50,
    "length": 1,
    "line": 4,
    "tokenModifiers": undefined,
    "tokenType": 5,
  },
]
`;
